### PR TITLE
fix(client): setup separate SyncArbiter for ViewClientActor with 4 threads

### PR DESCRIPTION
**Pull request by mikhailOK**
Change from #2752 + allow storage failures in code reachable from view client.

Test plan
---------
Run existing tests

### CODE REVIEW

File: test-utils/testlib/src/actix_utils.rs [hunk 1]
```diff
@@ -32,6 +34,8 @@ impl ShutdownableThread {
 impl Drop for ShutdownableThread {
     fn drop(&mut self) {
         self.shutdown();
+        // Leaving some time for all threads to stop after system is stopped.
+        thread::sleep(Duration::from_millis(100));
```
Comment (frol) [inline test-utils/testlib/src/actix_utils.rs:38]: It seems that it is not a solution, but a symptomatic patch. Is there an issue about this?

File: tests/test_tps_regression.rs [hunk 2]
```diff
@@ -140,17 +140,14 @@ mod test {
                                 .map(|idx| node.read().unwrap().user().get_block(idx).unwrap())
                                 .collect::<Vec<_>>();
                             for b in &blocks {
-                                let gas_used = b.chunks.iter().fold(0, |acc, chunk| {
+                                let _gas_used = b.chunks.iter().fold(0, |acc, chunk| {
```
Comment (frol) [inline tests/test_tps_regression.rs:143]: If this value is not used (why?), should we remove the whole computation?

File: neard/src/main.rs [hunk 3]
```diff
@@ -222,8 +222,9 @@ fn main() {
             }
 
             let system = System::new("NEAR");
-            start_with_config(home_dir, near_config);
+            let (_, _, arbiters) = start_with_config(home_dir, near_config);
             system.run().unwrap();
+            arbiters.into_iter().map(|mut a| a.join()).for_each(drop);
```
Comment (frol) [inline neard/src/main.rs:227]:
```suggestion
            arbiters.into_iter().for_each(|mut a| a.join());
```

File: neard/tests/rpc_nodes.rs [hunk 4]
```diff
@@ -594,6 +594,7 @@ fn test_get_execution_outcome(is_tx_successful: bool) {
         .start();
 
         system.run().unwrap();
+        clients.into_iter().flat_map(|c| c.2.into_iter().map(|mut a| a.join())).for_each(drop);
```
Comment (frol) [inline neard/tests/rpc_nodes.rs:597]: map().for_each(drop) can be replaced with simple for_each everywhere as far as I can tell

File: chain/client/src/view_client.rs [hunk 5]
```diff
@@ -911,3 +945,31 @@ impl Handler<GetGasPrice> for ViewClientActor {
...
+pub fn start_view_client(
+    validator_account_id: Option<AccountId>,
+    chain_genesis: ChainGenesis,
+    runtime_adapter: Arc<dyn RuntimeAdapter>,
+    network_adapter: Arc<dyn NetworkAdapter>,
+    config: ClientConfig,
+) -> Addr<ViewClientActor> {
+    let request_manager = Arc::new(RwLock::new(ViewClientRequestManager::new()));
+    SyncArbiter::start(config.view_client_threads, move || {
+        // ViewClientActor::start_in_arbiter(&Arbiter::current(), move |_ctx| {
```
Comment (bowenwang1996) [inline chain/client/src/view_client.rs:959]: should this line be removed?

File: neard/src/lib.rs [hunk 6]
```diff
@@ -171,5 +166,5 @@ pub fn start_with_config(
 
     trace!(target: "diagnostic", key="log", "Starting NEAR node with diagnostic activated");
 
-    (client_actor, view_client)
+    (client_actor, view_client, vec![client_arbiter, arbiter])
```
Comment (bowenwang1996) [inline neard/src/lib.rs:169]: why do we need to keep the arbiters?

File: test-utils/testlib/src/actix_utils.rs (hunk 1 above)
Comment (nearmax) [reply test-utils/testlib/src/actix_utils.rs]: I agree we should be explicitly waiting for all threads to stop.

**Comment by mikhailOK** [comment]
does not fix, only makes it affect chain store

File: neard/src/lib.rs (hunk 6 above)
Comment (mikhailOK) [reply neard/src/lib.rs:169]: we join them at the end, it seems like System::run blocks until stop signal and we want to wait for all threads to end, but not 100% sure it's needed

**Comment by mikhailOK** [comment]
Network spawns many threads since #2772
Regular client runs in one thread until we make the bigger storage change
JsonRpc requests that go to view client can now fail with these random storage errors, but we don't know yet how common they will be in practice

**Comment by nearmax** [comment]
Our JSONRpc server (not to be confused with ViewClienActor) still uses system arbiter, is it correct? If it is so then it means we have one thread for all RPC and regular ClientActor. CC @frol 

> JsonRpc requests that go to view client can now fail with these random storage errors, but we don't know yet how common they will be in practice

This is quite dangerous. Can we make sure our ViewClientActor does not serve torn data? If it is the case then lots of tools built on top of it will start failing, e.g. the bridge.

**Comment by SkidanovAlex** [comment]
> This is quite dangerous. Can we make sure our ViewClientActor does not serve torn data? If it is the case then lots of tools built on top of it will start failing, e.g. the bridge.

Our database is generally set up in a way that we do not modify stuff, we only insert and delete (except for very few exceptions). Correspondingly, the ViewClient will either serve your request successfully (all the data it expected was present throughout), or will fail.

Mostly failures will be due to concurrent GC, in which case the same request sent in ~1 second will also fail anyway.

The alternatives to this fix are:

1. Implement proper snapshot isolation level. Mikhail is working on it, but will not be done on the timeframe for Phase 1
2. Keep ViewClientActor with ClientActor on the same thread. That makes it very easy to practically stall block production, even if we disable JsonRPC (via e.g. state sync requests or transactions propagation).

Let's observe how ViewClient actually behaves with this change. In expectation, we will not observe any transient errors.

**Comment by nearmax** [comment]
> Correspondingly, the ViewClient will either serve your request successfully (all the data it expected was present throughout), or will fail.

Since we have multiple columns in the database can it be possible that while block production or block import is happening and we have populated only some of the columns the ViewClient will swoop in and read fresh data from some columns and old data from other columns?

> Let's observe how ViewClient actually behaves with this change. In expectation, we will not observe any transient errors.

I would prefer if we were certain before merging in. I am launching a persistent bridge today, and some of our partners will be using it. It would not look nice on us if it breaks because one of our RPC returns garbage data occasionally.

**Comment by SkidanovAlex** [comment]
> Since we have multiple columns in the database can it be possible that while block production or block import is happening and we have populated only some of the columns the ViewClient will swoop in and read fresh data from some columns and old data from other columns?

@mikhailOK can correct me if I'm wrong, but I believe that all the affected end points in the ViewClient have the following pattern:

1. Read A from the storage (say a block). If it doesn't exist, error out
2. Read B from the storage (say a chunk) that *we know exists given A exists*

This pattern can fail for two reasons:

a) We GCed B after A was read. Note that GC would not GC B before it GCed A, but naturally the timing could have been:

VC reads A -> GC deletes A -> GC deleted B -> VC reads B.

In this case the ViewClient will fail with this change (and would not fail before this change), but it is OK, since A is GCed, such a request is expected to start failing.

b) We inserted A before ViewClient read it, but haven't had a chance to insert B. Indeed, if the block production was not writing atomically, it would be possible:

BP writes A -> VC reads A -> VC attempts to read B -> BP writes B

I see you above mention that we do not use Transaction API, and have a concern that it might imply a possibility for dirty reads. Fortunately, in RocksDB batched writes are atomic, so dirty reads are impossible.


Note that even with atomic writes, it is still possible to have:

VC reads A -> (atomically BP writes A, BP writes B)

But it's OK, because VC will fail on reading A, and will not attempt to read B, so this data race is also acceptable.

**Comment by pmnoxx** [comment]
> SyncArbiter seems to be the right thing to use here, but does it mean that every other actor in our node runs on system arbiter, including the network and the RPC server? This would mean we are using a single OS thread for everything, but view clients. CC @pmnoxx
> 
> I'll delegate reviewing concurrency in this code to @pmnoxx and @frol
> 
> Also, @damons mentioned that our storage might allow dirty reads (we are using batched operations, instead of transactions API in rocksdb) so this PR might expose this issue now. If he is right, prepare for nodes returning garbage through RPC. CC @ailisp .

Http server starts it's own thread for accepting connections and workers.
PeerManagerActor start in arbiters, which uses it's own thread, and each Peer also has it's own dedicated Arbiter.
As far as I see, everything else runs on the main thread.

There is a bug in Actix library. It looks like Actix doesn't guarantee fairness. While there are messages in mailbox, Actix will prioritize processing messages until mailbox is empty. Execution of any other actors or tasks scheduled with run_later will be delayed.

**Comment by bowenwang1996** [comment]
@mikhailOK is there something that stops us from merging this PR?

**Comment by nearmax** [comment]
Thanks @pmnoxx . Since @pmnoxx  and @frol approved it, you don't need my approval.

**Comment by nearmax** [comment]
Actix unfairness sounds like a WAI. AFAIR run_later is designed to not give guarantees on when it is executed, it only guarantees that it is executed after the given time interval, this is from the documentation:
> Execute closure after specified period of time within same Actor and Context

Maksym Zavershynskyi
NEAR Protocol
https://near.org/

**Pull request merged**